libcloud/compute/drivers/vsphere.py

```diff
@@ -0,0 +1,1992 @@
+                    clusters.append(location)
+        if ex_show_hosts_in_drs:
+            hosts = hosts_all
+        else:
+            hosts_filter = [host for cluster in clusters
+                            for host in cluster.host]
+            hosts = [host for host in hosts_all if host not in hosts_filter]
+
+        for cluster in clusters:
+            locations.append(self._to_location(cluster))
+        for host in hosts:
+            locations.append(self._to_location(host))
+        return locations
+
+    def _to_location(self, data):
+        try:
+            if isinstance(data, vim.HostSystem):
+                extra = {
+                    "type": "host",
+                    "state": data.runtime.connectionState,
+                    "hypervisor": data.config.product.fullName,
+                    "vendor": data.hardware.systemInfo.vendor,
+                    "model": data.hardware.systemInfo.model,
+                    "ram": data.hardware.memorySize,
+                    "cpu": {
+                        "packages": data.hardware.cpuInfo.numCpuPackages,
+                        "cores": data.hardware.cpuInfo.numCpuCores,
+                        "threads": data.hardware.cpuInfo.numCpuThreads,
+                    },
+                    "uptime": data.summary.quickStats.uptime,
+                    "parent": str(data.parent)
+                }
+            elif isinstance(data, vim.ClusterComputeResource):
+                extra = {
+                    "type": "cluster",
+                    "overallStatus": data.overallStatus,
+                    "drs": data.configuration.drsConfig.enabled,
+                    'hosts': [host.name for host in data.host],
+                    'parent': str(data.parent)
+                }
+        except AttributeError as exc:
+            logger.error('Cannot convert location %s: %r' % (data.name, exc))
+            extra = {}
+        return NodeLocation(id=data.name, name=data.name, country=None,
+                            extra=extra, driver=self)
+
+    def ex_list_networks(self):
+        """
+        List networks
+        """
+        content = self.connection.RetrieveContent()
+        networks = content.viewManager.CreateContainerView(
+            content.rootFolder,
+            [vim.Network],
+            recursive=True
+        ).view
+
+        return [self._to_network(network) for network in networks]
+
+    def _to_network(self, data):
+        summary = data.summary
+        extra = {
+            'hosts': [h.name for h in data.host],
+            'ip_pool_name': summary.ipPoolName,
+            'ip_pool_id': summary.ipPoolId,
+            'accessible': summary.accessible
+        }
+        return VSphereNetwork(id=data.name, name=data.name, extra=extra)
+
+    def list_sizes(self):
+        """
+        Returns sizes
+        """
+        return []
+
+    def list_images(self, location=None, folder_ids=[]):
```

Comment:
Please don't default to a mutable value and do ``folder_ids=None`` and then inside the method ``folder_ids = folder_ids or []``.